lib/taxonomy.py

```diff
@@ -0,0 +1,58 @@
+import networkx as nx
+
+class Taxonomy:
+    """A class for operations on taxonomy"""
+
+    def __init__(self, graph):
+        self.graph = graph
+
+    @classmethod
+    def from_data(cls, data):
+        """Create data for the data present in full.json file
+        """
+        graph = nx.DiGraph()
+        nodes = []
+        edges = []
+        for elem in data:
+            nodes.append(elem)
+            # we prefer to have child relation, so we put parent first
+            edges.extend((parent, elem) for parent in data[elem]['parents'])
+        graph.add_nodes_from(nodes)
+        graph.add_edges_from(edges)
+        return cls(graph)
+
+    def ancestors(self, cat):
+        return nx.ancestors(self.graph, cat)
+
+    def descendants(self, cat):
+        return nx.descendants(self.graph, cat)
+    
+    def cats_complete(self, cats):
+        """Add ancestors categories to a list of cats
+        """
+        scats = set(cats)
+        for cat in cats:
+            scats.update(self.ancestors(cat))
+        return scats
+
+    def cats_compat(self, cats):
+        """Define the compatible and incompatible categories for a list of categories
+
+        The idea is that we don't want to penalize categories that are found by the model,
+        because it goes deeper than the
```

Comment:
some words are missing. Maybe you meant "we don't want to penalize deeper categories found by the model"

Reply:
well spotted

```
        The idea is that we don't want to penalize categories that are found by the model,
        because it goes deeper than what the dataset is providing
        (database is not necessarily complete)
```